subsys/bluetooth/host/long_work.h

```diff
@@ -0,0 +1,20 @@
+/* long_work.h - Workqueue intended for long-running operations. */
+
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#if defined(CONFIG_BT_LONG_WQ)
+
+void bt_long_work_schedule(struct k_work_delayable *dwork, k_timeout_t timeout);
+void bt_long_work_reschedule(struct k_work_delayable *dwork, k_timeout_t timeout);
+void bt_long_work_submit(struct k_work *dwork);
+
+#else
+
+#define bt_long_work_schedule k_work_schedule
+#define bt_long_work_reschedule k_work_reschedule
+#define bt_long_work_submit k_work_submit
+
+#endif /* CONFIG_BT_LONG_WQ */
```

Comment:
I am still on the fence on this: 

To me it is odd to be able to call `bt_long_work_schedule` and then _not_ having it being schedule on the long workqueue, but rather the system workqueue.

Reply:
Well, at least the API doesn't explicitly spell out that it promises to schedule the work through a dedicated work queue. It just says that it's for scheduling work of a long duration. How that's implemented is a configurable detail. Btw, I think static inline functions are cleaner whenever possible (e.g. for type checking), so I'd prefer to see that instead of defines for the else-branch.

Reply:
It'd be nice to see some test results of the system work queue option on some low performance platforms (e.g. nRF51) - I have a feeling you'll instantly see bad behaviour on those.

Reply:
yep, I'll try to get my hands on one and do that next week.

Reply:
So I spent half the day trying to make security L2 work on a 51422, but it hung in the middle of the procedure. That was with the long workqueue enabled, I didn't dare to try with the syswq.

It didn't even work on (today's) main, got an assert in the ull
```
uart:~$ ASSERTION FAIL [!radio_is_ready()] @ WEST_TOPDIR/zephyr/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:384
```

So I added a `depends on` in the ECC Kconfig, so the functionality/footprint stays the same (ECC runs in a low-prio thread).
Of course I'm not opposed to someone making it possible to run in the syswq, but I'd like that to be part of another PR.

I also addressed the other comments.

Reply:
> Btw, I think static inline functions are cleaner whenever possible (e.g. for type checking), so I'd prefer to see that instead of defines for the else-branch.

I don't disagree with that, but my suggestion is to remove the `#else` part all together. 

It's not a public API, so I guess it's fine as is, but I would prefer to return linker errors if `CONFIG_BT_LONG_WQ=n` and someone attempted to call `bt_long_work_submit`

Reply:
It's just that if we don't do this, then we have to litter the code with yet other ifdefs. 
I don't know if it's just me, but ifdefs add a cognitive load when trying to read code, as they break the code indentation. 
So I prefer that they are used only for big code blocks and not around individual function calls if possible.

If I use `IS_ENABLED` in an if-else, would the compiler optimize that like and #if-else block ?

Reply:
> If I use IS_ENABLED in an if-else, would the compiler optimize that like an #if-else block ?

Yes - IS_ENABLED will still compile it, but won't link it (and thus the final result is the same as if it wasn't compiled), which is why it is also the preferred method. 

It isn't a huge issue to keep as is, but I think it'll be more explicit what is going on if we don't silent forward the work item to the system workqueue instead if `CONFIG_BT_LONG_WQ=n`.